README.md

```diff
@@ -250,6 +165,114 @@ async fn main() {
+    // Print root directory.
+    println!("{:#?}", root_dir);
+}
+```
+
+You may notice that we store the `root_dir` returned by the `mkdir` operation, not the `dir` we started with. That is because WNFS internal state is immutable and every operation potentially returns a new root directory. This allows us to track and rollback changes when needed. It also makes collaborative editing easier to implement and reason about. You can find more examples in the [`crates/fs/examples/`][wnfs-examples] folder. And there is a basic demo of the filesystem immutability [here][wnfs-graph-demo].
+
+The private filesystem, on the other hand, is a bit more involved. [Hash Array Mapped Trie (HAMT)][hamt-wiki] is used as the intermediate format of private file tree before it is persisted to the blockstore. HAMT helps us hide the hierarchy of the file tree.
+
+```rust
+use wnfs::{
+    private::PrivateForest, MemoryBlockStore, Namefilter, PrivateDirectory, PrivateOpResult,
+};
+
+use chrono::Utc;
+use rand::thread_rng;
+
+use std::rc::Rc;
+
+#[async_std::main]
+async fn main() {
+    // Create a memory-based blockstore.
+    let store = &mut MemoryBlockStore::default();
+
+    // A random number generator the private filesystem can use.
+    let rng = &mut thread_rng();
+
+    // Create HAMT intermediate data structure.
+    let hamt = Rc::new(PrivateForest::new());
+
+    // Create a new private directory.
+    let dir = Rc::new(PrivateDirectory::new(
+        Namefilter::default(),
+        Utc::now(),
+        rng,
+    ));
+
+    // Add a file to /pictures/cats directory.
+    let PrivateOpResult { root_dir, hamt, .. } = dir
+        .mkdir(
+            &["pictures".into(), "cats".into()],
+            true,
+            Utc::now(),
+            hamt,
+            store,
+            rng,
+        )
+        .await
+        .unwrap();
+
+    // Add a file to /pictures/dogs/billie.jpg file.
+    let PrivateOpResult { root_dir, hamt, .. } = root_dir
+        .write(
+            &["pictures".into(), "dogs".into(), "billie.jpeg".into()],
+            true,
+            Utc::now(),
+            b"hello world".to_vec(),
+            hamt,
+            store,
+            rng,
+        )
+        .await
+        .unwrap();
+
+    // List all files in /pictures directory.
+    let PrivateOpResult { result, .. } = root_dir
+        .ls(&["pictures".into()], true, hamt, store)
+        .await
+        .unwrap();
+
+    println!("Files in /pictures: {:#?}", result);
+}
+```
+
+Namefilters are currently how we represent the identity key of a node in the filesystem. They have nice properties, one of which is the ability to check if one node belongs to another. This is necessary in a filesystem where metadata like hierarchy needs to be hidden from observing agents. One notable caveat with namefilters is that they can only reliably store information of a file tree 47 levels deep or less so there is a plan to replace them with cryptographic accumlators in the near future.
```

Comment:
```suggestion
Namefilters are currently how we represent the identity key of a node in the filesystem. They have nice properties, one of which is the ability to check if one node belongs to another. This is necessary in a filesystem where metadata like hierarchy needs to be hidden from observing agents. One notable caveat with namefilters is that they can only reliably store information of a file tree 47 levels deep or less so there is a plan to replace them with other cryptographic accumlators in the near future.
```

Technically namefilters are already cryptographic accumulators. But, really this change is a technicality.